### PR TITLE
[kolla] Update MySQL galera.cnf permissions to mysql and bind-address =  *

### PR DESCRIPTION
**Pull request by zerodayz**
Makes /etc/my.cnf.d/galera.cnf available in mariadb-openstack pod.

### CODE REVIEW

File: templates/mariadb/bin/config.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
         {
             "source": "/var/lib/config-data/galera.cnf",
             "dest": "/etc/my.cnf.d/galera.cnf",
-            "owner": "root",
+            "owner": "mysql",
```
Comment (olliewalsh) [inline templates/mariadb/bin/config.json:7]: Do we need this in init_config.json too?

**Comment by zerodayz** [comment]
@olliewalsh Sorry for the delay, I have few notes that I wanted to discuss before we start fixing it, because I believe we need some more clarification, especially on `mariadb_init.sh` file. Adding @dprince as there were changes from him I believe he may know.

## Wrong permissions on galera.cnf

We set incorrect permissions on `galera.cnf` hence the mysql user cannot read the file.

## Bind address is set to 127.0.0.0

This only works because the `mysqld` service is listening on * by default. Once we change the permissions, we would break `mysqld`.

So we are changing it to `*` which is both IPv4 and IPv6

## Wrong mariadb_init.sh

Log from init contaienr and more info later:

```bash
 oc logs openstack-db-init-xnkfm -f
+ sudo -E kolla_set_configs
sudo: unable to send audit message: Operation not permitted
INFO:__main__:Loading config file at /var/lib/kolla/config_files/config.json
INFO:__main__:Validating config file
INFO:__main__:Kolla config strategy set to: COPY_ALWAYS
INFO:__main__:Copying service configuration files
INFO:__main__:Deleting /etc/my.cnf.d/galera.cnf
INFO:__main__:Copying /var/lib/config-data/galera.cnf to /etc/my.cnf.d/galera.cnf
INFO:__main__:Setting permission for /etc/my.cnf.d/galera.cnf
INFO:__main__:Copying /var/lib/config-data/mariadb_init.sh to /usr/local/bin/mariadb_init.sh
INFO:__main__:Setting permission for /usr/local/bin/mariadb_init.sh
INFO:__main__:Writing out command to execute
INFO:__main__:Setting permission for /var/lib/mysql
++ cat /run_command
+ CMD=/usr/local/bin/mariadb_init.sh
+ ARGS=
+ [[ ! -n '' ]]
+ . kolla_extend_start
++ [[ -n 0 ]]
++ mysql_install_db
Installing MariaDB/MySQL system tables in '/var/lib/mysql' ...
OK

To start mysqld at boot time you have to copy
support-files/mysql.server to the right place for your system


Two all-privilege accounts were created.
One is root@localhost, it has no password, but you need to
be system 'root' user to connect. Use, for example, sudo mysql
The second is root@localhost, it has no password either, but
you need to be the system 'root' user to connect.
After connecting you can set the password, if you would need to be
able to connect as any of these users with a password and without sudo

See the MariaDB Knowledgebase at https://mariadb.com/kb

You can start the MariaDB daemon with:
cd '/usr' ; /usr/bin/mysqld_safe --datadir='/var/lib/mysql'

You can test the MariaDB daemon with mysql-test-run.pl
cd '/usr/mysql-test' ; perl mysql-test-run.pl

Please report any problems at https://mariadb.org/jira

The latest information about MariaDB is available at https://mariadb.org/.

Consider joining MariaDB's strong and vibrant community:
https://mariadb.org/get-involved/

++ bootstrap_db
++ TIMEOUT=60
++ [[ ! -S /var/lib/mysql/mysql.sock ]]
++ [[ ! -S /var/run/mysqld/mysqld.sock ]]
++ [[ 60 -gt 0 ]]
++ let TIMEOUT-=1
++ sleep 1
++ mysqld_safe --wsrep-new-cluster --skip-networking --wsrep-on=OFF --pid-file=/var/lib/mysql/mariadb.pid
220823 23:34:26 mysqld_safe Logging to '/var/lib/mysql/openstack-db-init-xnkfm.err'.
220823 23:34:26 mysqld_safe Starting mariadbd daemon with databases from /var/lib/mysql
++ [[ ! -S /var/lib/mysql/mysql.sock ]]
++ [[ ! -f /var/lib/mysql/mariadb.pid ]]
++ sudo -E kolla_security_reset
sudo: unable to send audit message: Operation not permitted
spawn mysql_secure_installation

NOTE: RUNNING ALL PARTS OF THIS SCRIPT IS RECOMMENDED FOR ALL MariaDB
      SERVERS IN PRODUCTION USE!  PLEASE READ EACH STEP CAREFULLY!

In order to log into MariaDB to secure it, we'll need the current
password for the root user. If you've just installed MariaDB, and
haven't set the root password yet, you should just press enter here.

Enter current password for root (enter for none):
OK, successfully used password, moving on...

Setting the root password or using the unix_socket ensures that nobody
can log into the MariaDB root user without the proper authorisation.

You already have your root account protected, so you can safely answer 'n'.

Switch to unix_socket authentication [Y/n] n
 ... skipping.

You already have your root account protected, so you can safely answer 'n'.

Change the root password? [Y/n] y
New password:
Re-enter new password:
Password updated successfully!
Reloading privilege tables..
 ... Success!


By default, a MariaDB installation has an anonymous user, allowing anyone
to log into MariaDB without having to have a user account created for
them.  This is intended only for testing, and to make the installation
go a bit smoother.  You should remove them before moving into a
production environment.

Remove anonymous users? [Y/n] y
 ... Success!

Normally, root should only be allowed to connect from 'localhost'.  This
ensures that someone cannot guess at the root password from the network.

Disallow root login remotely? [Y/n] n
 ... skipping.

By default, MariaDB comes with a database named 'test' that anyone can
access.  This is also intended only for testing, and should be removed
before moving into a production environment.

Remove test database and access to it? [Y/n] y
 - Dropping test database...
 ... Success!
 - Removing privileges on test database...
 ... Success!

Reloading the privilege tables will ensure that all changes made so far
will take effect immediately.

Reload privilege tables now? [Y/n] y
 ... Success!

Cleaning up...

All done!  If you've completed all of the above steps, your MariaDB
installation should now be secure.

Thanks for using MariaDB!
++ mysql -u root --password=12345678 -e 'GRANT ALL PRIVILEGES ON *.* TO '\''root'\''@'\''localhost'\'' IDENTIFIED BY '\''12345678'\'' WITH GRANT OPTION;'
++ mysql -u root --password=12345678 -e 'GRANT ALL PRIVILEGES ON *.* TO '\''root'\''@'\''%'\'' IDENTIFIED BY '\''12345678'\'' WITH GRANT OPTION;'
++ mysqladmin -uroot -p12345678 shutdown
++ exit 0
```

Looks run_command:

```bash
++ cat /run_command
+ CMD=/usr/local/bin/mariadb_init.sh
+ ARGS=
+ [[ ! -n '' ]]
+ . kolla_extend_start
++ [[ -n 0 ]]
++ mysql_install_db
```

And `. kolla_extend_start` runs `mysql_install_db` and `bootstrap_db` if `KOLLA_BOOTSTRAP=TRUE`

```bash
# This catches all cases of the BOOTSTRAP variable being set, including empty
if [[ "${!KOLLA_BOOTSTRAP[@]}" ]]; then
    mysql_install_db
    bootstrap_db
    exit 0
fi
```

```bash
sh-5.1$ cat /usr/local/bin/kolla_extend_start
#!/bin/bash

function bootstrap_db {
    mysqld_safe --wsrep-new-cluster --skip-networking --wsrep-on=OFF --pid-file=/var/lib/mysql/mariadb.pid &
    # Wait for the mariadb server to be "Ready" before starting the security reset with a max timeout
    # NOTE(huikang): the location of mysql's socket file varies depending on the OS distributions.
    # Querying the cluster status has to be executed after the existence of mysql.sock and mariadb.pid.
    TIMEOUT=${DB_MAX_TIMEOUT:-60}
    while [[ ! -S /var/lib/mysql/mysql.sock ]] && \
          [[ ! -S /var/run/mysqld/mysqld.sock ]] || \
          [[ ! -f /var/lib/mysql/mariadb.pid ]]; do
        if [[ ${TIMEOUT} -gt 0 ]]; then
            let TIMEOUT-=1
            sleep 1
        else
            exit 1
        fi
    done

    sudo -E kolla_security_reset
    mysql -u root --password="${DB_ROOT_PASSWORD}" -e "GRANT ALL PRIVILEGES ON *.* TO 'root'@'localhost' IDENTIFIED BY '${DB_ROOT_PASSWORD}' WITH GRANT OPTION;"
    mysql -u root --password="${DB_ROOT_PASSWORD}" -e "GRANT ALL PRIVILEGES ON *.* TO 'root'@'%' IDENTIFIED BY '${DB_ROOT_PASSWORD}' WITH GRANT OPTION;"
    mysqladmin -uroot -p"${DB_ROOT_PASSWORD}" shutdown
}

# This catches all cases of the BOOTSTRAP variable being set, including empty
if [[ "${!KOLLA_BOOTSTRAP[@]}" ]]; then
    mysql_install_db
    bootstrap_db
    exit 0
fi

if [[ "${!BOOTSTRAP_ARGS[@]}" ]]; then
    ARGS="${BOOTSTRAP_ARGS}"
fi
```

Which matches the logs from init container:

```bash
Thanks for using MariaDB!
++ mysql -u root --password=12345678 -e 'GRANT ALL PRIVILEGES ON *.* TO '\''root'\''@'\''localhost'\'' IDENTIFIED BY '\''12345678'\'' WITH GRANT OPTION;'
++ mysql -u root --password=12345678 -e 'GRANT ALL PRIVILEGES ON *.* TO '\''root'\''@'\''%'\'' IDENTIFIED BY '\''12345678'\'' WITH GRANT OPTION;'
++ mysqladmin -uroot -p12345678 shutdown
++ exit 0
```

So we are NOT actually running /usr/local/bin/mariadb_init.sh, but the above.

If we actually run the mariadb_init.sh as intended, it would likely fail. Look at the content of `mariadb_init.sh` is:

```bash
$ oc rsh openstack-db-init-nrjnh
sh-5.1$ cat /usr/local/bin/mariadb_init.sh
if [ -e /var/lib/mysql/mysql ]; then exit 0; fi
#echo -e "\n[mysqld]\nwsrep_provider=none" >> /etc/my.cnf
#kolla_set_configs
#sudo -u mysql -E kolla_extend_start
mkdir -p /var/lib/mysql
mysql_install_db
mysqld_safe --skip-networking --wsrep-on=OFF &
timeout <no value> /bin/bash -c "until mysqladmin -uroot -p'$DB_ROOT_PASSWORD' ping 2>/dev/null; do sleep 1; done"
mysql -uroot -p"$DB_ROOT_PASSWORD" -e "CREATE USER 'mysql'@'localhost';"
mysql -uroot -p"<no value>" -e "REVOKE ALL PRIVILEGES, GRANT OPTION FROM 'mysql'@'localhost';"
timeout <no value> mysqladmin -uroot -p"$DB_ROOT_PASSWORD" shutdown
```

Which as you can see is wrong for multiple reasons:

1. We no longer pass {{ .RootPassword }} to the container https://github.com/openstack-k8s-operators/mariadb-operator/commit/d882ed426e3164bd7c0cdd68a78c1dbdd35c161d
2. We no longer pass {{ .DBMaxTimeout }} to the container 
3. The value of both is therefore `<no value>`
4. The value of timeout should be set to `DB_MAX_TIMEOUT` https://github.com/openstack-k8s-operators/mariadb-operator/blob/d882ed426e3164bd7c0cdd68a78c1dbdd35c161d/pkg/dbinit.go#L39

@dprince  as I see many changes comes from you here, would you be able to comment regarding the above? Do we want to bootstrap the DB with root as we currently do and get rid of custom `mariadb_init.sh` ?

**Comment by zerodayz** [comment]
@olliewalsh Sorry for the delay, I have few notes that I wanted to discuss before we start fixing it, because I believe we need some more clarification, especially on `mariadb_init.sh` file. Adding @dprince as there were changes from him I believe he may know.

## Wrong permissions on galera.cnf

We set incorrect permissions on `galera.cnf` hence the mysql user cannot read the file.

## Bind address is set to 127.0.0.0

This only works because the `mysqld` service is listening on * by default. Once we change the permissions, we would break `mysqld`.

So we are changing it to `*` which is both IPv4 and IPv6

## Wrong mariadb_init.sh

Log from init contaienr and more info later:

```bash
 oc logs openstack-db-init-xnkfm -f
+ sudo -E kolla_set_configs
sudo: unable to send audit message: Operation not permitted
INFO:__main__:Loading config file at /var/lib/kolla/config_files/config.json
INFO:__main__:Validating config file
INFO:__main__:Kolla config strategy set to: COPY_ALWAYS
INFO:__main__:Copying service configuration files
INFO:__main__:Deleting /etc/my.cnf.d/galera.cnf
INFO:__main__:Copying /var/lib/config-data/galera.cnf to /etc/my.cnf.d/galera.cnf
INFO:__main__:Setting permission for /etc/my.cnf.d/galera.cnf
INFO:__main__:Copying /var/lib/config-data/mariadb_init.sh to /usr/local/bin/mariadb_init.sh
INFO:__main__:Setting permission for /usr/local/bin/mariadb_init.sh
INFO:__main__:Writing out command to execute
INFO:__main__:Setting permission for /var/lib/mysql
++ cat /run_command
+ CMD=/usr/local/bin/mariadb_init.sh
+ ARGS=
+ [[ ! -n '' ]]
+ . kolla_extend_start
++ [[ -n 0 ]]
++ mysql_install_db
Installing MariaDB/MySQL system tables in '/var/lib/mysql' ...
OK

To start mysqld at boot time you have to copy
support-files/mysql.server to the right place for your system


Two all-privilege accounts were created.
One is root@localhost, it has no password, but you need to
be system 'root' user to connect. Use, for example, sudo mysql
The second is root@localhost, it has no password either, but
you need to be the system 'root' user to connect.
After connecting you can set the password, if you would need to be
able to connect as any of these users with a password and without sudo

See the MariaDB Knowledgebase at https://mariadb.com/kb

You can start the MariaDB daemon with:
cd '/usr' ; /usr/bin/mysqld_safe --datadir='/var/lib/mysql'

You can test the MariaDB daemon with mysql-test-run.pl
cd '/usr/mysql-test' ; perl mysql-test-run.pl

Please report any problems at https://mariadb.org/jira

The latest information about MariaDB is available at https://mariadb.org/.

Consider joining MariaDB's strong and vibrant community:
https://mariadb.org/get-involved/

++ bootstrap_db
++ TIMEOUT=60
++ [[ ! -S /var/lib/mysql/mysql.sock ]]
++ [[ ! -S /var/run/mysqld/mysqld.sock ]]
++ [[ 60 -gt 0 ]]
++ let TIMEOUT-=1
++ sleep 1
++ mysqld_safe --wsrep-new-cluster --skip-networking --wsrep-on=OFF --pid-file=/var/lib/mysql/mariadb.pid
220823 23:34:26 mysqld_safe Logging to '/var/lib/mysql/openstack-db-init-xnkfm.err'.
220823 23:34:26 mysqld_safe Starting mariadbd daemon with databases from /var/lib/mysql
++ [[ ! -S /var/lib/mysql/mysql.sock ]]
++ [[ ! -f /var/lib/mysql/mariadb.pid ]]
++ sudo -E kolla_security_reset
sudo: unable to send audit message: Operation not permitted
spawn mysql_secure_installation

NOTE: RUNNING ALL PARTS OF THIS SCRIPT IS RECOMMENDED FOR ALL MariaDB
      SERVERS IN PRODUCTION USE!  PLEASE READ EACH STEP CAREFULLY!

In order to log into MariaDB to secure it, we'll need the current
password for the root user. If you've just installed MariaDB, and
haven't set the root password yet, you should just press enter here.

Enter current password for root (enter for none):
OK, successfully used password, moving on...

Setting the root password or using the unix_socket ensures that nobody
can log into the MariaDB root user without the proper authorisation.

You already have your root account protected, so you can safely answer 'n'.

Switch to unix_socket authentication [Y/n] n
 ... skipping.

You already have your root account protected, so you can safely answer 'n'.

Change the root password? [Y/n] y
New password:
Re-enter new password:
Password updated successfully!
Reloading privilege tables..
 ... Success!


By default, a MariaDB installation has an anonymous user, allowing anyone
to log into MariaDB without having to have a user account created for
them.  This is intended only for testing, and to make the installation
go a bit smoother.  You should remove them before moving into a
production environment.

Remove anonymous users? [Y/n] y
 ... Success!

Normally, root should only be allowed to connect from 'localhost'.  This
ensures that someone cannot guess at the root password from the network.

Disallow root login remotely? [Y/n] n
 ... skipping.

By default, MariaDB comes with a database named 'test' that anyone can
access.  This is also intended only for testing, and should be removed
before moving into a production environment.

Remove test database and access to it? [Y/n] y
 - Dropping test database...
 ... Success!
 - Removing privileges on test database...
 ... Success!

Reloading the privilege tables will ensure that all changes made so far
will take effect immediately.

Reload privilege tables now? [Y/n] y
 ... Success!

Cleaning up...

All done!  If you've completed all of the above steps, your MariaDB
installation should now be secure.

Thanks for using MariaDB!
++ mysql -u root --password=12345678 -e 'GRANT ALL PRIVILEGES ON *.* TO '\''root'\''@'\''localhost'\'' IDENTIFIED BY '\''12345678'\'' WITH GRANT OPTION;'
++ mysql -u root --password=12345678 -e 'GRANT ALL PRIVILEGES ON *.* TO '\''root'\''@'\''%'\'' IDENTIFIED BY '\''12345678'\'' WITH GRANT OPTION;'
++ mysqladmin -uroot -p12345678 shutdown
++ exit 0
```

Looks run_command:

```bash
++ cat /run_command
+ CMD=/usr/local/bin/mariadb_init.sh
+ ARGS=
+ [[ ! -n '' ]]
+ . kolla_extend_start
++ [[ -n 0 ]]
++ mysql_install_db
```

And `. kolla_extend_start` runs `mysql_install_db` and `bootstrap_db` if `KOLLA_BOOTSTRAP=TRUE`

```bash
# This catches all cases of the BOOTSTRAP variable being set, including empty
if [[ "${!KOLLA_BOOTSTRAP[@]}" ]]; then
    mysql_install_db
    bootstrap_db
    exit 0
fi
```

```bash
sh-5.1$ cat /usr/local/bin/kolla_extend_start
#!/bin/bash

function bootstrap_db {
    mysqld_safe --wsrep-new-cluster --skip-networking --wsrep-on=OFF --pid-file=/var/lib/mysql/mariadb.pid &
    # Wait for the mariadb server to be "Ready" before starting the security reset with a max timeout
    # NOTE(huikang): the location of mysql's socket file varies depending on the OS distributions.
    # Querying the cluster status has to be executed after the existence of mysql.sock and mariadb.pid.
    TIMEOUT=${DB_MAX_TIMEOUT:-60}
    while [[ ! -S /var/lib/mysql/mysql.sock ]] && \
          [[ ! -S /var/run/mysqld/mysqld.sock ]] || \
          [[ ! -f /var/lib/mysql/mariadb.pid ]]; do
        if [[ ${TIMEOUT} -gt 0 ]]; then
            let TIMEOUT-=1
            sleep 1
        else
            exit 1
        fi
    done

    sudo -E kolla_security_reset
    mysql -u root --password="${DB_ROOT_PASSWORD}" -e "GRANT ALL PRIVILEGES ON *.* TO 'root'@'localhost' IDENTIFIED BY '${DB_ROOT_PASSWORD}' WITH GRANT OPTION;"
    mysql -u root --password="${DB_ROOT_PASSWORD}" -e "GRANT ALL PRIVILEGES ON *.* TO 'root'@'%' IDENTIFIED BY '${DB_ROOT_PASSWORD}' WITH GRANT OPTION;"
    mysqladmin -uroot -p"${DB_ROOT_PASSWORD}" shutdown
}

# This catches all cases of the BOOTSTRAP variable being set, including empty
if [[ "${!KOLLA_BOOTSTRAP[@]}" ]]; then
    mysql_install_db
    bootstrap_db
    exit 0
fi

if [[ "${!BOOTSTRAP_ARGS[@]}" ]]; then
    ARGS="${BOOTSTRAP_ARGS}"
fi
```

Which matches the logs from init container:

```bash
Thanks for using MariaDB!
++ mysql -u root --password=12345678 -e 'GRANT ALL PRIVILEGES ON *.* TO '\''root'\''@'\''localhost'\'' IDENTIFIED BY '\''12345678'\'' WITH GRANT OPTION;'
++ mysql -u root --password=12345678 -e 'GRANT ALL PRIVILEGES ON *.* TO '\''root'\''@'\''%'\'' IDENTIFIED BY '\''12345678'\'' WITH GRANT OPTION;'
++ mysqladmin -uroot -p12345678 shutdown
++ exit 0
```

So we are NOT actually running /usr/local/bin/mariadb_init.sh, but the above.

If we actually run the mariadb_init.sh as intended, it would likely fail. Look at the content of `mariadb_init.sh` is:

```bash
$ oc rsh openstack-db-init-nrjnh
sh-5.1$ cat /usr/local/bin/mariadb_init.sh
if [ -e /var/lib/mysql/mysql ]; then exit 0; fi
#echo -e "\n[mysqld]\nwsrep_provider=none" >> /etc/my.cnf
#kolla_set_configs
#sudo -u mysql -E kolla_extend_start
mkdir -p /var/lib/mysql
mysql_install_db
mysqld_safe --skip-networking --wsrep-on=OFF &
timeout <no value> /bin/bash -c "until mysqladmin -uroot -p'$DB_ROOT_PASSWORD' ping 2>/dev/null; do sleep 1; done"
mysql -uroot -p"$DB_ROOT_PASSWORD" -e "CREATE USER 'mysql'@'localhost';"
mysql -uroot -p"<no value>" -e "REVOKE ALL PRIVILEGES, GRANT OPTION FROM 'mysql'@'localhost';"
timeout <no value> mysqladmin -uroot -p"$DB_ROOT_PASSWORD" shutdown
```

Which as you can see is wrong for multiple reasons:

1. We no longer pass {{ .RootPassword }} to the container https://github.com/openstack-k8s-operators/mariadb-operator/commit/d882ed426e3164bd7c0cdd68a78c1dbdd35c161d
2. We no longer pass {{ .DBMaxTimeout }} to the container 
3. The value of both is therefore `<no value>`
4. The value of timeout should be set to `DB_MAX_TIMEOUT` https://github.com/openstack-k8s-operators/mariadb-operator/blob/d882ed426e3164bd7c0cdd68a78c1dbdd35c161d/pkg/dbinit.go#L39

@dprince  as I see many changes comes from you here, would you be able to comment regarding the above? Do we want to bootstrap the DB with root as we currently do and get rid of custom `mariadb_init.sh` ?

**Comment by dprince** [comment]
@zerodayz 

Hey, nice assessment on this. The code here has been somewhat ignored recently with the expectation that we are going to rewrite the guts to use Galera instead. Whether this occurs by continuing to use this operator codebase or copying the useful bits out of this repository (perhaps the MariaDBDatabase stuff) into a new galera-operator codebase has yet to be decided. What is here is certainly not production ready (it uses a naive pod instead of a deployment) but it does functionally work.

The history of this is I think we started from a YAML that looked something like this: https://github.com/dprince/k8s-yamls/blob/master/mariadb/mariadb.yaml

And then we created an operator, and then switched it over to use Kolla bootstrap instead (which it should have done to begin with but I can't remember why I wrote mariadb_init.sh long ago).

Your analysis is sound however and I think we can ditch mariadb_init.sh entirely.

As for further work here perhaps lets sync with Damien tomorrow (his morning, your afternoon evening) and see what a best plan of action is to integrate the Galera work he has been doing. Before we can integrate that work however it does need to meet at a minimum what this operator already supports with things like full OLM support (ClusterServiceVersion correct, tested, working, etc) and the equivalent of the MariaDBDatabase CR to assist in creating the permissions for each openstack service to use the database as a shared resource.

**Pull request closed without merging**